dev mode in multimodule fails to restart server

hello,

Please let me report an issue regarding the liberty-maven-plugin.

When i try to restart a server configured in a submodule, using the 'r' key in the console, it usually fails as the plugin tries to add some features in the running server instance, which is killed as the features might be incompatible. 

- mvn -am -pl submodule-ws liberty:dev : launch the server which starts normally without errors
- the liberty plugin is set in the submodule-ws pom.xml with the following configuration : 
```xml
<groupId>io.openliberty.tools</groupId>
<artifactId>liberty-maven-plugin</artifactId>
<version>3.8.2</version>
<configuration>
	<assemblyArtifact>
	<groupId>com.ibm.websphere.appserver.runtime</groupId>
	<artifactId>wlp-javaee7</artifactId>
	<version>23.0.0.8</version>
</assemblyArtifact>
</configuration>
<executions>
	<execution>
	     <id>install-feature</id>
	     <phase>prepare-package</phase>
	     <goals>
		 <goal>install-feature</goal>
	    </goals>
		<configuration>
			<features>
				<acceptLicense>true</acceptLicense>
				<feature>jsp-2.2</feature>
				<feature>localConnector-1.0</feature>
				<feature>cdi-1.0</feature>
				<feature>jaxrs-1.1</feature>
				<feature>jpa-2.0</feature>
				<feature>jaxws-2.2</feature>
				<feature>ejbLite-3.1</feature>
				<feature>beanValidation-1.0</feature>
				<feature>transportSecurity-1.0</feature>
				<feature>jdbc-4.2</feature>
			</features>
		</configuration>
	</execution>
	<execution>
		<id>create</id>
		<phase>initialize</phase>
		<goals>
			<goal>install-server</goal>
		</goals>
	</execution>
</executions>
</plugin>
```
- server.xml has the following features activated (matchs the pom) : 
```xml
<server description="sampleapp">
    <featureManager>
        <feature>jsp-2.2</feature>
        <feature>localConnector-1.0</feature>
        <feature>cdi-1.0</feature>
        <feature>jaxrs-1.1</feature>
        <feature>jpa-2.0</feature>
        <feature>jaxws-2.2</feature>
        <feature>ejbLite-3.1</feature>
        <feature>beanValidation-1.0</feature>
        <feature>transportSecurity-1.0</feature>
        <feature>jdbc-4.2</feature>
    </featureManager>
```
When i try to restart i have the following error : 

[INFO] Parsing the server file for features and includes: defaultServer/server.xml
[INFO] plugin listed esa: []
[INFO] Installing features: [ejbLite-3.1, jsp-2.2, javaee-7.0, cdi-1.0, transportSecurity-1.0, jaxws-2.2, localConnector-1.0, jaxrs-1.1, jpa-2.0, beanValidation-1.0]

[ERROR] Could not restart the server.
io.openliberty.tools.common.plugins.util.PluginExecutionException: org.apache.maven.plugin.MojoExecutionException: Error installing features for server defaultServer
    at io.openliberty.tools.maven.server.DevMojo$DevMojoUtil.libertyInstallFeature (DevMojo.java:408)
    at io.openliberty.tools.common.plugins.util.DevUtil.restartServer (DevUtil.java:1853)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.readInput (DevUtil.java:2610)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.run (DevUtil.java:2577)
    at java.lang.Thread.run (Thread.java:750)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error installing features for server defaultServer
    at io.openliberty.tools.maven.server.InstallFeatureMojo.doInstallFeatures (InstallFeatureMojo.java:77)
    at io.openliberty.tools.maven.server.InstallFeatureMojo.execute (InstallFeatureMojo.java:66)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:120)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojo (StartDebugMojoSupport.java:362)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojoInstallFeature (StartDebugMojoSupport.java:347)
    at io.openliberty.tools.maven.server.DevMojo.runLibertyMojoInstallFeature (DevMojo.java:1870)
    at io.openliberty.tools.maven.server.DevMojo$DevMojoUtil.libertyInstallFeature (DevMojo.java:406)
    at io.openliberty.tools.common.plugins.util.DevUtil.restartServer (DevUtil.java:1853)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.readInput (DevUtil.java:2610)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.run (DevUtil.java:2577)
    at java.lang.Thread.run (Thread.java:750)
Caused by: io.openliberty.tools.common.plugins.util.PluginExecutionException: A feature conflict error occurred while installing features: [ejbLite-3.1, jsp-2.2, javaee-7.0, cdi-1.0, transportSecurity-1.0, jaxws-2.2, localConnector-1.0, jaxrs-1.1, jpa-2.0, beanValidation-1.0]: CWWKF1405E: Les fonctions de singleton jsp-2.2 et jsp-2.3 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton cdi-1.0 et cdi-1.2 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton beanValidation-1.0 et beanValidation-1.1 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton eeCompatible-6.0,eeCompatible-7.0, et eeCompatible-8.0 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton ejb-3.1 et ejb-3.2 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton interceptor-1.1 et interceptor-1.2 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton jpa-2.0 et jpa-2.1 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton internal-1.6 et internal-1.7 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton el-2.2 et el-3.0 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton jaxrs-1.1 et jaxrs-2.0 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton transaction-1.1 et transaction-1.2 sont incompatibles et ne peuvent pas être installées simultanément.
CWWKF1405E: Les fonctions de singleton ejbLite-3.1 et ejbLite-3.2 sont incompatibles et ne peuvent pas être installées simultanément.

    at io.openliberty.tools.common.plugins.util.InstallFeatureUtil.installFeatures (InstallFeatureUtil.java:720)
    at io.openliberty.tools.maven.server.InstallFeatureMojo.installFeatures (InstallFeatureMojo.java:112)
    at io.openliberty.tools.maven.server.InstallFeatureMojo.doInstallFeatures (InstallFeatureMojo.java:75)
    at io.openliberty.tools.maven.server.InstallFeatureMojo.execute (InstallFeatureMojo.java:66)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:120)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojo (StartDebugMojoSupport.java:362)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojoInstallFeature (StartDebugMojoSupport.java:347)
    at io.openliberty.tools.maven.server.DevMojo.runLibertyMojoInstallFeature (DevMojo.java:1870)
    at io.openliberty.tools.maven.server.DevMojo$DevMojoUtil.libertyInstallFeature (DevMojo.java:406)
    at io.openliberty.tools.common.plugins.util.DevUtil.restartServer (DevUtil.java:1853)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.readInput (DevUtil.java:2610)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.run (DevUtil.java:2577)
    at java.lang.Thread.run (Thread.java:750)

It seems the plugin looses track of the actual server configuration at this point.

This pb doesnt show when invoking dev mode within the -ws submodule directory.

Regards
André

hello @cherylking,
Yes, sorry i did not give you a feedback on the suggested workaround.
It doesn't work as well, it fails also but with a different error.

```
[ERROR] Could not restart the server.
io.openliberty.tools.common.plugins.util.PluginExecutionException: org.apache.maven.plugin.MojoExecutionException: CWWKM2173E: Failed to install application from project project:lib:jar:0.0.1-SNAPSHOT. The project packaging type is not supported.
    at io.openliberty.tools.maven.server.DevMojo$DevMojoUtil.libertyDeploy (DevMojo.java:417)
    at io.openliberty.tools.common.plugins.util.DevUtil.restartServer (DevUtil.java:1855)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.readInput (DevUtil.java:2610)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.run (DevUtil.java:2577)
    at java.lang.Thread.run (Thread.java:833)
Caused by: org.apache.maven.plugin.MojoExecutionException: CWWKM2173E: Failed to install application from project project:lib:jar:0.0.1-SNAPSHOT. The project packaging type is not supported.
    at io.openliberty.tools.maven.applications.DeployMojo.installProject (DeployMojo.java:204)
    at io.openliberty.tools.maven.applications.DeployMojo.doDeploy (DeployMojo.java:103)
    at io.openliberty.tools.maven.applications.DeployMojo.execute (DeployMojo.java:61)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:120)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojo (StartDebugMojoSupport.java:362)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojoDeploy (StartDebugMojoSupport.java:328)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojoDeploy (StartDebugMojoSupport.java:311)
    at io.openliberty.tools.maven.server.DevMojo$DevMojoUtil.libertyDeploy (DevMojo.java:415)
    at io.openliberty.tools.common.plugins.util.DevUtil.restartServer (DevUtil.java:1855)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.readInput (DevUtil.java:2610)
    at io.openliberty.tools.common.plugins.util.DevUtil$HotkeyReader.run (DevUtil.java:2577)
    at java.lang.Thread.run (Thread.java:833)
```
André

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/StartDebugMojoSupport.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/StartDebugMojoSupport.java
@@ -288,9 +288,13 @@ public abstract class StartDebugMojoSupport extends ServerFeatureSupport {
         }
         Plugin projectPlugin = currentProject.getPlugin(LIBERTY_MAVEN_PLUGIN_GROUP_ID + ":" + LIBERTY_MAVEN_PLUGIN_ARTIFACT_ID);
         if (projectPlugin == null) {
+            getLog().debug("Did not find liberty-maven-plugin configured in currentProject: "+currentProject.toString());
             projectPlugin = getPluginFromPluginManagement(LIBERTY_MAVEN_PLUGIN_GROUP_ID, LIBERTY_MAVEN_PLUGIN_ARTIFACT_ID, currentProject);
+        } else {
+            getLog().debug("Found liberty-maven-plugin configured in currentProject: "+currentProject.toString());
         }
         if (projectPlugin == null) {
+            getLog().debug("Did not find liberty-maven-plugin in pluginManagement in currentProject: "+currentProject.toString());
             projectPlugin = plugin(LIBERTY_MAVEN_PLUGIN_GROUP_ID, LIBERTY_MAVEN_PLUGIN_ARTIFACT_ID, "LATEST");
         }
         if (version != null) {
@@ -370,8 +374,10 @@ public abstract class StartDebugMojoSupport extends ServerFeatureSupport {
     private void runLibertyMojo(String goal, Xpp3Dom config) throws MojoExecutionException {
         getLog().info("Running liberty:" + goal);
         getLog().debug("configuration:\n" + config);
-        executeMojo(getLibertyPlugin(), goal(goal), config,
-                executionEnvironment(project, session, pluginManager));
+        getLog().debug("project:\n" + project);
+        MavenSession tempSession = session.clone();
+        tempSession.setCurrentProject(project);
+        executeMojo(getLibertyPlugin(), goal(goal), config, executionEnvironment(project, tempSession, pluginManager));
     }
 
     private void copyDependencies() throws MojoExecutionException, IOException {

